n2v/grid/grider.py

```diff
@@ -0,0 +1,918 @@
+            z = psi4.core.Vector.from_array(grid[2][idx:])
+            if if_w:
+                w = psi4.core.Vector.from_array(grid[3][idx:])
+            else:
+                w = psi4.core.Vector.from_array(np.zeros_like(grid[2][idx:]))  # When w is not necessary and not given
+            blocks.append(psi4.core.BlockOPoints(x, y, z, w, extens))
+
+        max_functions = 0 if 0 > len(blocks[-1].functions_local_to_global()) \
+                                      else len(blocks[-1].functions_local_to_global())
+
+        if self.ref == 1:
+            point_func = psi4.core.RKSFunctions(self.basis, max_points, max_functions)
+        else:
+            point_func = psi4.core.UKSFunctions(self.basis, max_points, max_functions)
+
+        return blocks, npoints, point_func
+
+    def generate_mesh(self, grid=default_grid):
+        """
+        Genrates Mesh from 3 separate linear spaces
+        needed for cubic grid.
+
+        Parameters
+        ----------
+        grid: np.ndarray
+            grid needed to be turned into 3d array
+        
+        Returns
+        -------
+        out_grid: np.ndarray
+            Mesh grid as 3dimensional array
+        """
+
+        shape = (grid.shape[1], grid.shape[1], grid.shape[1])
+        X,Y,Z = np.meshgrid(grid[0],grid[1],grid[2])
+        X = X.reshape((X.shape[0] * X.shape[1] * X.shape[2], 1))
+        Y = Y.reshape((Y.shape[0] * Y.shape[1] * Y.shape[2], 1))
+        Z = Z.reshape((Z.shape[0] * Z.shape[1] * Z.shape[2], 1))
+        grid = np.concatenate((X,Y,Z), axis=1).T
+
+        return grid, shape
+
+    def generate_dft_grid(self, wfn):
+        """
+        Extracts weights from wfn object
+
+        Parameters
+        ----------
+        wfn: psi4.core.{RKS, UKS}
+        """
+
+        try:
+            vpot = wfn.V_potential()
+        except:
+            raise ValueError("Wfn object does not contain a Vpotential object. Try with one from a DFT calculation")
+
+        nblocks = vpot.nblocks()
+        blocks = [vpot.get_block(i) for i in range(nblocks)]
+        npoints = vpot.grid().npoints()
+        points_function = vpot.properties()[0]
+
+        x = np.empty((npoints))
+        y = np.empty((npoints))
+        z = np.empty((npoints))
+        w = np.empty((npoints))
+
+        offset = 0
+        for i_block in blocks:
+            b_points = i_block.npoints()
+            offset += b_points
+
+            x[offset - b_points : offset] = i_block.x().np
+            y[offset - b_points : offset] = i_block.x().np
+            z[offset - b_points : offset] = i_block.x().np
+            w[offset - b_points : offset] = i_block.x().np
+
+        return [x,y,z,w]
+
+
+    #Quantities on Grid
+    def on_grid_ao(self, matrix, grid=None, vpot=None, cubic_grid=False):
+        """
+        Generates a quantity on the grid given its ao representation
+
+        Parameters
+        ----------
+        matrix: np.ndarray
+            Quantity on ao basis
+        grid: np.ndarray
```

Comment:
Always specify the shape of array. Here it should be (3, total_num_grid_points) or (4, total_num_grid_points)